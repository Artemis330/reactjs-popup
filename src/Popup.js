import React from "react";
import calculatePosition from "./Utils";
// https://philipwalton.com/articles/what-no-one-told-you-about-z-index/

import styles from "./popup.css.js";

export default class Popup extends React.Component {
  static defaultProps = {
    closeOnDocumentClick: false,
    defaultOpen: false,
    triggerOn: "click",
    contentStyle: {},
    arrowStyle: {},
    overlayStyle: {},
    className: "",
    position: "bottom center",
    modal: false,
    arrow: true,
    children: () => <span> Your Content Here !!</span>
  };
  state = {
    isOpen: this.props.defaultOpen
  };
  timeOut = 0;

  constructor(props) {
    super(props);
    this.setTriggerRef = r => (this.TriggerEl = r);
    this.setContentRef = r => (this.ContentEl = r);
    this.setArrowRef = r => (this.ArrowEl = r);
    this.setHelperRef = r => (this.HelperEl = r);
  }

  componentDidMount() {
    if (this.props.defaultOpen) this.setPosition();
  }

  togglePopup = () => {
    this.setState(
      prevState => ({
        isOpen: !prevState.isOpen
      }),
      () => this.state.isOpen && this.setPosition()
    );
  };
  openPopup = () => {
    this.setState({ isOpen: true }, () => this.setPosition());
  };
  closePopup = () => {
    this.setState({ isOpen: false });
  };
  onMouseEnter = () => {
    clearTimeout(this.timeOut);
    this.openPopup();
  };
  onMouseLeave = () => {
    clearTimeout(this.timeOut);
    this.timeOut = setTimeout(() => this.closePopup(), 300);
  };

  setPosition = () => {
    const { modal, arrow, position } = this.props;
    if (modal) return;
    const helper = this.HelperEl.getBoundingClientRect();
    const trigger = this.TriggerEl.getBoundingClientRect();
    const content = this.ContentEl.getBoundingClientRect();
    const cords = calculatePosition(trigger, content, position, arrow);
    this.ContentEl.style.top = cords.top - helper.top + "px";
    this.ContentEl.style.left = cords.left - helper.left + "px";
    if (arrow) {
      this.ArrowEl.style.transform = cords.transform;
      this.ArrowEl.style.top = cords.arrowTop;
      this.ArrowEl.style.left = cords.arrowLeft;
    }
    if (
      this.TriggerEl.style.position == "static" ||
      this.TriggerEl.style.position == ""
    )
      this.TriggerEl.style.position = "relative";
  };

  addWarperAction = () => {
    const { contentStyle, className, modal, triggerOn } = this.props;
    const popupContentStyle = modal
      ? styles.popupContent.modal
      : styles.popupContent.tooltip;

    const childrenElementProps = {
      className: `popup-content ${className}`,
      style: Object.assign({}, popupContentStyle, contentStyle),
      ref: this.setContentRef,
      onClick: e => {
        e.stopPropagation();
      }
    };
    if (!modal && triggerOn === "hover") {
      childrenElementProps.onMouseEnter = this.onMouseEnter;
      childrenElementProps.onMouseLeave = this.onMouseLeave;
    }
    return childrenElementProps;
  };
  renderTrigger = () => {
    const triggerProps = {};
    const { triggerOn } = this.props;
    triggerProps.ref = this.setTriggerRef;
    switch (triggerOn) {
      case "click":
        triggerProps.onClick = this.togglePopup;
        break;
      case "hover":
        triggerProps.onMouseEnter = this.onMouseEnter;
        triggerProps.onMouseLeave = this.onMouseLeave;
        break;
    }
    return React.cloneElement(this.props.trigger, triggerProps);
  };

  renderContent = () => {
    const { arrow, modal, arrowStyle } = this.props;
    return (
      <div {...this.addWarperAction()}>
        {arrow &&
          !modal && (
            <div
              ref={this.setArrowRef}
              style={Object.assign({}, styles.popupArrow, arrowStyle)}
            />
          )}
        {typeof this.props.children === "function"
          ? this.props.children(this.state.isOpen, this.closePopup)
          : this.props.children}
      </div>
    );
  };

  render() {
    const { modal, overlayStyle } = this.props;
    const ovStyle = modal ? styles.overlay.modal : styles.overlay.tooltip;

    return [
      <div
        key="H"
        style={{ position: "absolute", top: "0px", left: "0px" }}
        ref={this.setHelperRef}
      />,
      this.state.isOpen && (
        <div
          key="O"
          className="popup-overlay"
          style={Object.assign({}, ovStyle, overlayStyle)}
          onClick={
            this.props.closeOnDocumentClick ? this.closePopup : undefined
          }
        >
          {modal && this.renderContent()}
        </div>
      ),

      this.state.isOpen && !modal && this.renderContent(),
      this.renderTrigger()
    ];
  }
}

if (process.env.NODE_ENV !== "production") {
  const PropTypes = require("prop-types");

  Popup.propTypes = {
    arrowStyle: PropTypes.object,
    contentStyle: PropTypes.object,
    overlayStyle: PropTypes.object,
    className: PropTypes.string,
    modal: PropTypes.bool,
    closeOnDocumentClick: PropTypes.bool,
    trigger: PropTypes.element.isRequired,
    triggerOn: PropTypes.oneOf(["hover", "click"]),
    children: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.element,
      PropTypes.string
    ]).isRequired,
    position: PropTypes.oneOf([
      "top left",
      "top center",
      "top right",
      "bottom left",
      "bottom center",
      "bottom right",
      "right top",
      "right center",
      "right bottom",
      "left top",
      "left center",
      "left bottom"
    ])
  };
}
