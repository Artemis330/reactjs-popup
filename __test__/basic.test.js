import React from "react";
import { shallow, mount } from "enzyme";
import Enzyme from "enzyme";
import Popup from "../src/Popup";
import { shallowToJson } from "enzyme-to-json";

import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({ adapter: new Adapter() });

const PopupTest = props => (
  <Popup {...props} trigger={<button> Trigger</button>}>
    popup content
  </Popup>
);

test("it should render correctly ", () => {
  const popup = shallow(<PopupTest triggerOn="click" />);
  expect(shallowToJson(popup)).toMatchSnapshot();
});
test("it should render correctly on click  ", () => {
  const popup = mount(<PopupTest />);
  popup.find("button").simulate("click");
  expect(shallowToJson(popup)).toMatchSnapshot();
});

test("it should render correctly on click (triggerOn = 'click') ", () => {
  const popup = mount(<PopupTest triggerOn="click" />);
  popup.find("button").simulate("click");
  expect(shallowToJson(popup)).toMatchSnapshot();
});

test("it should render correctly on hover (triggerOn = 'hover') ", () => {
  const popup = mount(<PopupTest triggerOn="hover" />);
  popup.find("button").simulate("mouseEnter");
  expect(shallowToJson(popup)).toMatchSnapshot();
});

test("it should render popup content correctly on click ", () => {
  const popup = mount(<PopupTest triggerOn="click" />);
  popup.find("button").simulate("click");
  expect(popup.find("div.popup-content").text()).toBe("popup content");
});

test("it should render popup content correctly on hover ", () => {
  const popup = mount(<PopupTest triggerOn="hover" />);
  popup.find("button").simulate("mouseEnter");
  expect(popup.find("div.popup-content").text()).toBe("popup content");
});

// closeOnDocumentClick Tests

test("it should close on click outside popup ", () => {
  const popup = mount(<PopupTest triggerOn="click" />);
  popup.find("button").simulate("click");
  expect(shallowToJson(popup)).toMatchSnapshot();
  popup.find("div.overlay").simulate("click");
  expect(shallowToJson(popup)).toMatchSnapshot();
});
test("it should close on click outside popup (closeOnDocumentClick = true )", () => {
  const popup = mount(
    <PopupTest triggerOn="click" closeOnDocumentClick={true} />
  );
  popup.find("button").simulate("click");
  expect(shallowToJson(popup)).toMatchSnapshot();
  popup.find("div.overlay").simulate("click");
  expect(shallowToJson(popup)).toMatchSnapshot();
});
