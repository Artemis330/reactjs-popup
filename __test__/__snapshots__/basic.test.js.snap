// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should close on click outside popup  1`] = `
<PopupTest
  triggerOn="click"
>
  <Popup
    className=""
    closeOnDocumentClick={false}
    defaultOpen={false}
    position="bottom,center"
    style={Object {}}
    trigger={
      <button>
         Trigger
      </button>
    }
    triggerOn="click"
  >
    <button
      onClick={[Function]}
    >
       Trigger
    </button>
    <div
      className="overlay"
    >
      <div
        className="popup-content  "
        onClick={[Function]}
        style={
          Object {
            "position": "fixed",
          }
        }
      >
        <div
          className="popup-arrow"
        />
        popup content
      </div>
    </div>
  </Popup>
</PopupTest>
`;

exports[`it should close on click outside popup  2`] = `
<PopupTest
  triggerOn="click"
>
  <Popup
    className=""
    closeOnDocumentClick={false}
    defaultOpen={false}
    position="bottom,center"
    style={Object {}}
    trigger={
      <button>
         Trigger
      </button>
    }
    triggerOn="click"
  >
    <button
      onClick={[Function]}
    >
       Trigger
    </button>
    <div
      className="overlay"
    >
      <div
        className="popup-content  "
        onClick={[Function]}
        style={
          Object {
            "position": "fixed",
          }
        }
      >
        <div
          className="popup-arrow"
        />
        popup content
      </div>
    </div>
  </Popup>
</PopupTest>
`;

exports[`it should close on click outside popup (closeOnDocumentClick = true ) 1`] = `
<PopupTest
  closeOnDocumentClick={true}
  triggerOn="click"
>
  <Popup
    className=""
    closeOnDocumentClick={true}
    defaultOpen={false}
    position="bottom,center"
    style={Object {}}
    trigger={
      <button>
         Trigger
      </button>
    }
    triggerOn="click"
  >
    <button
      onClick={[Function]}
    >
       Trigger
    </button>
    <div
      className="overlay"
      onClick={[Function]}
    >
      <div
        className="popup-content  "
        onClick={[Function]}
        style={
          Object {
            "position": "fixed",
          }
        }
      >
        <div
          className="popup-arrow"
        />
        popup content
      </div>
    </div>
  </Popup>
</PopupTest>
`;

exports[`it should close on click outside popup (closeOnDocumentClick = true ) 2`] = `
<PopupTest
  closeOnDocumentClick={true}
  triggerOn="click"
>
  <Popup
    className=""
    closeOnDocumentClick={true}
    defaultOpen={false}
    position="bottom,center"
    style={Object {}}
    trigger={
      <button>
         Trigger
      </button>
    }
    triggerOn="click"
  >
    <button
      onClick={[Function]}
    >
       Trigger
    </button>
  </Popup>
</PopupTest>
`;

exports[`it should render correctly  1`] = `
<Popup
  className=""
  closeOnDocumentClick={false}
  defaultOpen={false}
  position="bottom,center"
  style={Object {}}
  trigger={
    <button>
       Trigger
    </button>
  }
  triggerOn="click"
>
  popup content
</Popup>
`;

exports[`it should render correctly on click   1`] = `
<PopupTest>
  <Popup
    className=""
    closeOnDocumentClick={false}
    defaultOpen={false}
    position="bottom,center"
    style={Object {}}
    trigger={
      <button>
         Trigger
      </button>
    }
    triggerOn="click"
  >
    <button
      onClick={[Function]}
    >
       Trigger
    </button>
    <div
      className="overlay"
    >
      <div
        className="popup-content  "
        onClick={[Function]}
        style={
          Object {
            "position": "fixed",
          }
        }
      >
        <div
          className="popup-arrow"
        />
        popup content
      </div>
    </div>
  </Popup>
</PopupTest>
`;

exports[`it should render correctly on click (triggerOn = 'click')  1`] = `
<PopupTest
  triggerOn="click"
>
  <Popup
    className=""
    closeOnDocumentClick={false}
    defaultOpen={false}
    position="bottom,center"
    style={Object {}}
    trigger={
      <button>
         Trigger
      </button>
    }
    triggerOn="click"
  >
    <button
      onClick={[Function]}
    >
       Trigger
    </button>
    <div
      className="overlay"
    >
      <div
        className="popup-content  "
        onClick={[Function]}
        style={
          Object {
            "position": "fixed",
          }
        }
      >
        <div
          className="popup-arrow"
        />
        popup content
      </div>
    </div>
  </Popup>
</PopupTest>
`;

exports[`it should render correctly on hover (triggerOn = 'hover')  1`] = `
<PopupTest
  triggerOn="hover"
>
  <Popup
    className=""
    closeOnDocumentClick={false}
    defaultOpen={false}
    position="bottom,center"
    style={Object {}}
    trigger={
      <button>
         Trigger
      </button>
    }
    triggerOn="hover"
  >
    <button
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
       Trigger
    </button>
    <div
      className="overlay"
    >
      <div
        className="popup-content  "
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "position": "fixed",
          }
        }
      >
        <div
          className="popup-arrow"
        />
        popup content
      </div>
    </div>
  </Popup>
</PopupTest>
`;
